BinaryTree.java

```diff
@@ -0,0 +1,137 @@
+
+        if(left != "")
+            return left;
+        return right;
+    }
+
+    private BinaryTree lowestCommonAncestor(BinaryTree tree1, BinaryTree tree2) {
+        if(tree1 == null || tree2 == null)
+            return null;
+
+        BinaryTree leftRes = null;
+        BinaryTree rightRes = null;
+
+        if (this.left != null)
+            leftRes = this.left.lowestCommonAncestor(tree1, tree2);
+        if (this.right != null)
+            rightRes = this.right.lowestCommonAncestor(tree1, tree2);
+        if(this.left == null)
+            return rightRes;
+        if(this.right == null)
+            return leftRes;
+
+        if(this.left.contains(tree1) && this.right.contains(tree2))
+            return this;
+        if(this.left.contains(tree2) && this.right.contains(tree1))
+            return this;
+
+        if(leftRes != null)
+            return leftRes;
+        return rightRes;
+    }
+
+    private boolean contains(BinaryTree contained) {
+        if(this.info == contained.info) {
+           if (this.isIdentical(contained))
+               return true;
+        }
+
+        boolean containedLeft = false;
+        boolean containedRight = false;
+
+        if(this.left != null)
+            containedLeft = this.left.contains(contained);
+
+        if(this.right != null)
+            containedRight = this.right.contains(contained);
+
+        return (containedLeft || containedRight);
+    }
+
+    private boolean isIdentical(BinaryTree tree) {
+        if(this.info != tree.info)
+            return false;
+        if(this.left == null && tree.left != null)
+            return false;
+        if(this.left != null && tree.left == null)
+            return false;
+        if(this.right == null && tree.right != null)
+            return false;
+        if(this.right != null && tree.right == null)
+            return false;
+
+        boolean leftMatch = true;
+        boolean rightMatch = true;
+        if(this.left != null) {
+            leftMatch = this.left.isIdentical(tree.left);
+        }
+        if(this.right != null) {
+            rightMatch = this.right.isIdentical(tree.right);
+        }
+        return (leftMatch && rightMatch);
+    }
+
+    public static void main(String[] args) {
+        BinaryTree bt = new BinaryTree(16,
+                new BinaryTree(9,
+                        new BinaryTree(3,
+                                new BinaryTree(1, null, null),
+                                new BinaryTree(5, null, null)
+                        ),
+                        new BinaryTree(14, null, null)
+                ),
+                new BinaryTree(18,
+                        null,
+                        new BinaryTree(19, null, null)
+                )
+        );
+
+        // Question 1
+        System.out.println(bt.printAncestors(5));
+
+        // Question 2
+        System.out.println(bt.lowestCommonAncestor(new BinaryTree(5, null, null), new BinaryTree(14, null, null)).getInfo());
```

Comment:
Finding common ancestor has quite a few edge cases that need to be handled. This println only covers one case. Additionally it has to be then read by human and manually checked if printed value is correct, which is error prone.

Could you change the code so that lowestCommonAncestor is called with different values covering different edge cases (like key not present in the tree, one key being ancestor of the other etc) and for each of them prints status, whether returned value was correct or not?
For bonus points you could look at [JUnit](https://github.com/junit-team/junit4/wiki/Getting-started), but writing your own solution is perfectly acceptable.

Note, that testing code (including main() method) should not be inside BinaryTree class, as the only concern of BinaryTree should be implementation of binary tree and methods that operate on it. Testing code should live in a separate class, for example BinaryTreeTest or something similar.